Add authorization

yesoer opened the pull request:
resolves #40
resolves #51 

## Important changes/additions : 
In the jwt token, now we store the user id (if a compelling argument is made we can change it back, it just makes more sense to me).

User id from jwt token aswell as user type are now stored in fiber context locals for easy access in route handlers.

Also from [13ccb80](https://github.com/PPSKSY-Cluster/backend/pull/66/commits/13ccb80e15da1ee4c64095de1f9d4b6eff835d55) on where we should i.e. control when to return a PW (set it to empty string) or where we allow to set the user type : 
```
I think "what data part(s) do we want to add/get/modify/delete" is a question for handlers, 
not the db client. 
This way we keep the db more versatile and this behavior would't change for other dbs 
(=no reimplementation needed). 
Also for some things we will need authorization information, which we already have in our handlers 
(context locals) but would have to pass down for handling in db package
```

vDawgg commented on tests/c_resources_test.go:
```diff
@@ -17,9 +18,14 @@ func Test_cresources(t *testing.T) {
-	user := db.User{Username: "foo", Password: "bar"}
-	tokenStr, createdUser := createUserAndLogin(t, app, user)
+	// create user with admin rights
+	user := db.User{Username: "foo", Password: "bar", Type: db.AdminUT}
```
Adding an initial setup-step that does exactly this would also be a good idea. You would have to make it mandatory for the superAdmin to change the default credentials right after though.

vDawgg replied:
Yes, exactly that. Would be fine if not included in this PR but Ill open an issue on this

yesoer commented on api/users.go:
```diff
@@ -118,12 +120,36 @@ func userUpdateHandler() func(*fiber.Ctx) error {
 		}
 
+		// only super-admin or the user themself are allowed to edit
+		isSuperAdmin := c.Locals("jwtUserType") != db.SuperAdminUT
```
True that's wrong

vDawgg commented on api/users.go:
```diff
@@ -118,12 +120,36 @@ func userUpdateHandler() func(*fiber.Ctx) error {
+		// only super admins may change the user type and only the user may
+		// change their PW
+		if isSuperAdmin {
+			u.Password = ""
```
Wont the sA still be unable to change its own pw?

yesoer replied:
yes true will change that

vDawgg commented on api/users.go:
```diff
@@ -118,12 +120,36 @@ func userUpdateHandler() func(*fiber.Ctx) error {
 			return fiber.NewError(fiber.StatusNotFound, err.Error())
 		}
 
-		user.ID = id
```
The thing is, that ids arent returned correctly when not being reapplied to the object. If ommitted the json would contain ID=0000000000...

yesoer commented:
@vDawgg your corrections have been addressed. Wanna have another look at it ? I also rebased and adapted everything for the refresh token and refactored to fiber.NewError in all other places where I didn't use it yet

vDawgg commented on api/users.go:
```diff
@@ -10,7 +10,7 @@ import (
 	userRouter.Post("/", userCreateHandler())
 
-	userRouter.Use(auth.CheckToken())
+	userRouter.Use(auth.CheckToken(db.UserUT))
```
Why use this as the default call? I guess Im not 100% sure on how access is checked here, if the function is called once with a default

yesoer replied:
Not sure if I understand what exactly you mean but the way this works is just like the route handlers : CheckToken is a function that returns another function and said other function is the actual middleware that will be called when someone sends a request. And that actual middleware will compare the user from the jwt token to the type it was given on initalization

vDawgg replied:
Arent Admins then disallowed from using this route?

yesoer replied:
If we were to only check against the given type yes but CheckToken() uses userTypeIncludes(). So AdminUT works for UserUT restricted routes but not vice versa. There might even be a prettier way to do this i.e. if user types were numbers we could simply do ">=" checks but I don't want to do that to the frontend team right now.
